homeassistant/components/xiaomi_miio/__init__.py

```diff
@@ -1 +1,69 @@
 """Support for Xiaomi Miio."""
+import logging
+
+from homeassistant import config_entries, core
+from homeassistant.const import CONF_HOST, CONF_TOKEN
+from homeassistant.helpers import device_registry as dr
+
+from .config_flow import CONF_FLOW_TYPE, CONF_GATEWAY
+from .const import DOMAIN
+from .gateway import ConnectXiaomiGateway
+
+_LOGGER = logging.getLogger(__name__)
+
+GATEWAY_PLATFORMS = ["alarm_control_panel"]
+
+
+async def async_setup(hass: core.HomeAssistant, config: dict):
+    """Set up the Xiaomi Miio component."""
+    return True
+
+
+async def async_setup_entry(
+    hass: core.HomeAssistant, entry: config_entries.ConfigEntry
+):
+    """Set up the Xiaomi Miio components from a config entry."""
+    hass.data[DOMAIN] = {}
+    if entry.data[CONF_FLOW_TYPE] == CONF_GATEWAY:
+        if not await async_setup_gateway_entry(hass, entry):
+            return False
+
+    return True
+
+
+async def async_setup_gateway_entry(
+    hass: core.HomeAssistant, entry: config_entries.ConfigEntry
+):
+    """Set up the Xiaomi Gateway component from a config entry."""
+    host = entry.data[CONF_HOST]
+    token = entry.data[CONF_TOKEN]
+    name = entry.title
+    gateway_id = entry.data["gateway_id"]
+
+    # Connect to gateway
+    gateway = ConnectXiaomiGateway(hass)
+    if not await gateway.async_connect_gateway(host, token):
+        return False
+    gateway_info = gateway.gateway_info
+
+    hass.data[DOMAIN][entry.entry_id] = gateway.gateway_device
+
+    gateway_model = f"{gateway_info.model}-{gateway_info.hardware_version}"
```

Comment:
Why combine both model and hardware version here? Does model not carry enough information here? This function could be generalized to `async_setup_device_entry` to allow easier porting of other platforms to use the config flow in the future.

Reply:
I included the hardware_version since there is no specific device property in which I could place it, see https://developers.home-assistant.io/docs/device_registry_index.

It is nice for users to be able to easily find there hardware_version besides the model, because if a user is experiancing problems whith commands not working/firmware updates or other problems it is nice to compare hardware versions with other users and see if others with exactly the same device (model and hardware version) experiance the same issues.

Reply:
I agree that it'd would be useful to have the hardware information available (and it is not just specific for xiaomi_miio), but I would much rather prefer to have it fixed in the core. This will require an architecture issue first, though...

Reply:
Well that is definetly for some other PR.
I will leave it as it is and it can be changed later on if someone want to make an architecture issue